sdks/java/extensions/ml/src/main/java/org/apache/beam/sdk/extensions/ml/RecommendationAIPredict.java

```diff
@@ -0,0 +1,155 @@
+import com.google.cloud.recommendationengine.v1beta1.PredictResponse;
+import com.google.cloud.recommendationengine.v1beta1.PredictionServiceClient;
+import com.google.cloud.recommendationengine.v1beta1.UserEvent;
+import com.google.protobuf.util.JsonFormat;
+import java.io.IOException;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TupleTagList;
+import org.json.JSONObject;
+
+/**
+ * A {@link PTransform} using the Recommendations AI API (https://cloud.google.com/recommendations).
+ * Takes an input {@link PCollection} of {@link GenericJson}s and creates {@link
+ * PredictResponse.PredictionResult}s.
+ *
+ * <p>It is possible to provide a catalog name to which you want to add the user event (defaults to
+ * "default_catalog"). It is possible to provide a event store to which you want to add the user
+ * event (defaults to "default_event_store"). A placement id for the recommendation engine placement
+ * to be used.
+ */
+@AutoValue
+@SuppressWarnings({"nullness"})
+public abstract class RecommendationAIPredict
+    extends PTransform<PCollection<GenericJson>, PCollectionTuple> {
+
+  /** @return ID of Google Cloud project to be used for creating catalog items. */
+  public abstract @Nullable String projectId();
+
+  /** @return Name of the catalog where the catalog items will be created. */
+  public abstract @Nullable String catalogName();
+
+  /** @return Name of the event store where the user events will be created. */
+  public abstract @Nullable String eventStore();
+
+  /** @return ID of the recommendation engine placement. */
+  public abstract String placementId();
+
+  public static final TupleTag<PredictResponse.PredictionResult> SUCCESS_TAG =
+      new TupleTag<PredictResponse.PredictionResult>() {};
+
+  public static final TupleTag<UserEvent> FAILURE_TAG = new TupleTag<UserEvent>() {};
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+    /** @param projectId ID of Google Cloud project to be used for creating catalog items. */
+    public abstract Builder setProjectId(@Nullable String projectId);
+
+    /** @param catalogName Name of the catalog where the catalog items will be created. */
+    public abstract Builder setCatalogName(@Nullable String catalogName);
+
+    /** @param eventStore Name of the event store where the user events will be created. */
+    public abstract Builder setEventStore(@Nullable String eventStore);
+
+    /** @param placementId of the recommendation engine placement. */
+    public abstract Builder setPlacementId(String placementId);
+
+    public abstract RecommendationAIPredict build();
+  }
+
+  public static Builder newBuilder() {
```

Comment:
you can remove `public` from this builder so it can only be used within the package